imageio/core/imopen.py

```diff
@@ -0,0 +1,254 @@
+from numbers import Number
+import re
+import numpy as np
+
+from .format import FormatManager, MODENAMES
+from .request import Request, RETURN_BYTES
+from .util import Singleton
+
+
+class _imopen(metaclass=Singleton):
+    def __init__(self):
+        self._known_plugins = list()
+
+    def __call__(self, uri, *args, plugin=None, api='legacy', **kwargs):
+        if api == "legacy":
+            return LegacyPlugin(uri, *args, plugin=plugin, **kwargs)
+
+        if plugin is not None:
+            candidate_plugins = filter(
+                lambda x: x.name == plugin,
+                self._known_plugins
+            )
+
+            if len(candidate_plugins) == 0:
+                raise ValueError(
+                    f"'{plugin}' is not a registered plugin name.")
+
+            return candidate_plugins[0](uri, *args, **kwargs)
+
+        else:
+            candidate_plugins = filter(
+                lambda x: x.can_read(uri) and x.can_write(uri),
+                self._known_plugins
+            )
+
+            if len(candidate_plugins) > 0:
+                return candidate_plugins[0]
```

Comment:
A simple loop with exist ASAP instead of a filter is may be better since only the first occurrence is used.

Reply:
I agree. Would we want to do some sorting on the `candidate_plugins` though? Like allow users (or plugins) to specify a priority or `weight` for different `uri`s? (just a random thought)

Reply:
I have no strong opinion on that, plugins priority can be considered as a future enhancement to be implemented :wink:.